src/backend/distributed/commands/multi_copy.c

```diff
@@ -2923,3 +3070,380 @@ CopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist)
 	return attnums;
 	/* *INDENT-ON* */
 }
+
+
+/*
+ * CreateConnectionStateHash constructs a hash table which maps from socket
+ * number to CopyConnectionState, passing the provided MemoryContext to
+ * hash_create for hash allocations.
+ */
+static HTAB *
+CreateConnectionStateHash(MemoryContext memoryContext)
+{
+	HTAB *connectionStateHash = NULL;
+	int hashFlags = 0;
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(int);
+	info.entrysize = sizeof(CopyConnectionState);
+	info.hcxt = memoryContext;
+	hashFlags = (HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+
+	connectionStateHash = hash_create("Copy Connection State Hash", 128, &info,
+									  hashFlags);
+
+	return connectionStateHash;
+}
+
+
+/*
+ * CreateShardStateHash constructs a hash table which maps from shard
+ * identifier to CopyShardState, passing the provided MemoryContext to
+ * hash_create for hash allocations.
+ */
+static HTAB *
+CreateShardStateHash(MemoryContext memoryContext)
+{
+	HTAB *shardStateHash = NULL;
+	int hashFlags = 0;
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(uint64);
+	info.entrysize = sizeof(CopyShardState);
+	info.hcxt = memoryContext;
+	hashFlags = (HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+
+	shardStateHash = hash_create("Copy Shard State Hash", 128, &info, hashFlags);
+
+	return shardStateHash;
+}
+
+
+/*
+ * GetConnectionState finds existing CopyConnectionState for a connection in the
+ * provided hash. If not found, then a default structure is returned.
+ */
+static CopyConnectionState *
+GetConnectionState(HTAB *connectionStateHash, MultiConnection *connection)
+{
+	CopyConnectionState *connectionState = NULL;
+	bool found = false;
+	int socket = PQsocket(connection->pgConn);
```

Comment:
Minor: Maybe move this line inside `!found` block.

Reply:
is it possible that `PQsocket` returns -1 when a connection has failed?

Reply:
> Minor: Maybe move this line inside `!found` block.

This is used in the `hash_search()` call before the `!found` block.

Reply:
> is it possible that `PQsocket` returns -1 when a connection has failed?

Looking into this. If `GetPlacementListConnectionIfCached()` can return bad connections, then maybe, otherwise no. I will look into this.

Reply:
> is it possible that `PQsocket` returns -1 when a connection has failed?

Postgres docs say "This will not change during normal operation, but could change during connection setup or reset", so I think this shouldn't be -1 here. I added an assert.